[bert] Add the intra-op parallelism for equal operator

jianyuh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28810 [bert] Add the intra-op parallelism for equal operator**

Similar to https://github.com/pytorch/pytorch/pull/28464 and https://github.com/pytorch/pytorch/pull/28477, we would like to enable the intra-op parallelism for layer norm. This will be mapped to the parallel performance win for the BERT/RoBERTa model.

Benchmarking the RoBERTa model with 20 threads:

Before this Diff: P120104857
```
equal                    11.16%           305.851ms        11.16%           305.851ms        4.248ms          NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]     
```


After this Diff:
(grain size is the third parameter when using `at::parallel_for`. As measured below, the performance differences between these different grain size seem to be subtle.)

- grain size = `TH_OMP_OVERHEAD_THRESHOLD`:
```
equal                    1.43%            36.056ms         1.43%            36.056ms         500.783us        NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]
```

- grain size = `HYPER_TH_OMP_OVERHEAD_THRESHOLD`:
```
equal                    1.41%            35.126ms         1.41%            35.126ms         487.855us        NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]
```

- grain size = 1:
```
equal                    1.43%            35.632ms         1.43%            35.632ms         494.886us        NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]
```

Note that the size of  `HYPER_TH_OMP_OVERHEAD_THRESHOLD` and  `TH_OMP_OVERHEAD_THRESHOLD` can be found in https://github.com/pytorch/pytorch/blob/master/aten/src/TH/generic/THTensorApply.hpp#L7-L10 . `HYPER_TH_OMP_OVERHEAD_THRESHOLD` is used for more fine-grained tasks since HYPER_TH_OMP_OVERHEAD_THRESHOLD = TH_OMP_OVERHEAD_THRESHOLD / 16.


Differential Revision: [D18165752](https://our.internmc.facebook.com/intern/diff/D18165752/)

jianyuh committed:
[bert] Add the intra-op parallelism for equal operator

Similar to https://github.com/pytorch/pytorch/pull/28464, we would like to enable the intra-op parallelism for layer norm. This will be mapped to the parallel performance win for the BERT/RoBERTa model.

Differential Revision: [D18165752](https://our.internmc.facebook.com/intern/diff/D18165752/)

[ghstack-poisoned]

jspark1105 commented on aten/src/TH/generic/THTensorMoreMath.cpp:
```diff
@@ -28,9 +28,18 @@ int THTensor_(equal)(THTensor *ta, THTensor* tb)
+    at::parallel_for(
+        0,
+        sz,
+        HYPER_TH_OMP_OVERHEAD_THRESHOLD,
```
I *think* HYPER_TH_OMP_OVERHEAD_THRESHOLD is used for more expensive functions like cosh so it can generate too fine-grain tasks. Maybe TH_OMP_OVERHEAD_THRESHOLD is a right one to use here?

ilia-cher replied:
I'd suggest running a simple benchmark to tune this param

jianyuh replied:
The definition of `HYPER_TH_OMP_OVERHEAD_THRESHOLD` and `TH_OMP_OVERHEAD_THRESHOLD` can be found in https://github.com/pytorch/pytorch/blob/master/aten/src/TH/generic/THTensorApply.hpp#L7-L10 .
As you mentioned,  `HYPER_TH_OMP_OVERHEAD_THRESHOLD` is used for more fine-grained tasks since HYPER_TH_OMP_OVERHEAD_THRESHOLD = TH_OMP_OVERHEAD_THRESHOLD / 16.

However, as I measured, the performance differences between these different grain size are subtle:

(grain size is the third parameter when using `at::parallel_for`. As measured below, the performance differences between these different grain size seem to be subtle.)

- grain size = `TH_OMP_OVERHEAD_THRESHOLD`:
```
equal                    1.43%            36.056ms         1.43%            36.056ms         500.783us        NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]
```

- grain size = `HYPER_TH_OMP_OVERHEAD_THRESHOLD`:
```
equal                    1.41%            35.126ms         1.41%            35.126ms         487.855us        NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]
```

- grain size = 1:
```
equal                    1.43%            35.632ms         1.43%            35.632ms         494.886us        NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]
```

VitalyFedyunin commented on aten/src/TH/generic/THTensorMoreMath.cpp:
```diff
@@ -28,9 +28,18 @@ int THTensor_(equal)(THTensor *ta, THTensor* tb)
+        HYPER_TH_OMP_OVERHEAD_THRESHOLD,
+        [&](int64_t begin, int64_t end) {
+          for (auto iter = begin; iter < end; iter++) {
+            if (equal && tap[iter] != tbp[iter]) {
```
You probably want:
```cpp
if (!equal) { break; }
if (tap[iter] != tbp[iter]) {
	equal = 0;
    break;
}
```

jianyuh replied:
Thanks! Updated with your suggestion! The code should have the early exit from the loop when `equal` is 0 already (changed by other threads).

jianyuh committed:
Update on "[bert] Add the intra-op parallelism for equal operator"


Similar to https://github.com/pytorch/pytorch/pull/28464 and https://github.com/pytorch/pytorch/pull/28477, we would like to enable the intra-op parallelism for layer norm. This will be mapped to the parallel performance win for the BERT/RoBERTa model.

Differential Revision: [D18165752](https://our.internmc.facebook.com/intern/diff/D18165752/)

[ghstack-poisoned]

jianyuh commented:
TODO: more benchmarking to choose the granularity size.

jspark1105 commented on aten/src/TH/generic/THTensorMoreMath.cpp:
```diff
@@ -28,9 +28,21 @@ int THTensor_(equal)(THTensor *ta, THTensor* tb)
+        TH_OMP_OVERHEAD_THRESHOLD,
+        [&](int64_t begin, int64_t end) {
+          for (auto iter = begin; iter < end; iter++) {
+            if (equal) {
```
You wanted
if (!equal)
?

jianyuh replied:
Thanks! Sorry for correcting me! Shouldn't write the code at late night...

jianyuh committed:
Update on "[bert] Add the intra-op parallelism for equal operator"


Similar to https://github.com/pytorch/pytorch/pull/28464 and https://github.com/pytorch/pytorch/pull/28477, we would like to enable the intra-op parallelism for layer norm. This will be mapped to the parallel performance win for the BERT/RoBERTa model.

Differential Revision: [D18165752](https://our.internmc.facebook.com/intern/diff/D18165752/)

[ghstack-poisoned]

jianyuh commented:
Done the benchmarking. Updated the result in the summary.

ilia-cher commented on aten/src/TH/generic/THTensorMoreMath.cpp:
```diff
@@ -28,9 +28,21 @@ int THTensor_(equal)(THTensor *ta, THTensor* tb)
+        TH_OMP_OVERHEAD_THRESHOLD,
+        [&](int64_t begin, int64_t end) {
+          for (auto iter = begin; iter < end; iter++) {
+            if (!equal) {
```
I'm a bit concerned about read/writes, in some cases they will be atomic but I'm not sure this is always true;
could you use local to the scope equal variable and then write it into atomic int variable defined outside

VitalyFedyunin replied:
In case of non atomic write worst thing that could happen is making few more loop iterations. It is much cheaper than any synchronization.

ilia-cher replied:
what about read/write race?

ilia-cher replied:
using std::atomic<int> would probably be also negligibly different as it would just use atomic processor instructions

jianyuh replied:
Updated to `std::atomic<int>`.

jianyuh committed:
Update on "[bert] Add the intra-op parallelism for equal operator"


Similar to https://github.com/pytorch/pytorch/pull/28464 and https://github.com/pytorch/pytorch/pull/28477, we would like to enable the intra-op parallelism for layer norm. This will be mapped to the parallel performance win for the BERT/RoBERTa model.

Benchmarking the RoBERTa model with 20 threads:

Before this Diff: P120104857
```
equal                    11.16%           305.851ms        11.16%           305.851ms        4.248ms          NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]     
```


After this Diff:
(grain size is the third parameter when using `at::parallel_for`. As measured below, the performance differences between these different grain size seem to be subtle.)

- grain size = `TH_OMP_OVERHEAD_THRESHOLD`:
```
equal                    1.43%            36.056ms         1.43%            36.056ms         500.783us        NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]
```

- grain size = `HYPER_TH_OMP_OVERHEAD_THRESHOLD`:
```
equal                    1.41%            35.126ms         1.41%            35.126ms         487.855us        NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]
```

- grain size = 1:
```
equal                    1.43%            35.632ms         1.43%            35.632ms         494.886us        NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]
```

Note that the size of  `HYPER_TH_OMP_OVERHEAD_THRESHOLD` and  `TH_OMP_OVERHEAD_THRESHOLD` can be found in https://github.com/pytorch/pytorch/blob/master/aten/src/TH/generic/THTensorApply.hpp#L7-L10 . `HYPER_TH_OMP_OVERHEAD_THRESHOLD` is used for more fine-grained tasks since HYPER_TH_OMP_OVERHEAD_THRESHOLD = TH_OMP_OVERHEAD_THRESHOLD / 16.


Differential Revision: [D18165752](https://our.internmc.facebook.com/intern/diff/D18165752/)

[ghstack-poisoned]

jianyuh committed:
Update on "[bert] Add the intra-op parallelism for equal operator"


Similar to https://github.com/pytorch/pytorch/pull/28464 and https://github.com/pytorch/pytorch/pull/28477, we would like to enable the intra-op parallelism for layer norm. This will be mapped to the parallel performance win for the BERT/RoBERTa model.

Benchmarking the RoBERTa model with 20 threads:

Before this Diff: P120104857
```
equal                    11.16%           305.851ms        11.16%           305.851ms        4.248ms          NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]     
```


After this Diff:
(grain size is the third parameter when using `at::parallel_for`. As measured below, the performance differences between these different grain size seem to be subtle.)

- grain size = `TH_OMP_OVERHEAD_THRESHOLD`:
```
equal                    1.43%            36.056ms         1.43%            36.056ms         500.783us        NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]
```

- grain size = `HYPER_TH_OMP_OVERHEAD_THRESHOLD`:
```
equal                    1.41%            35.126ms         1.41%            35.126ms         487.855us        NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]
```

- grain size = 1:
```
equal                    1.43%            35.632ms         1.43%            35.632ms         494.886us        NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]
```

Note that the size of  `HYPER_TH_OMP_OVERHEAD_THRESHOLD` and  `TH_OMP_OVERHEAD_THRESHOLD` can be found in https://github.com/pytorch/pytorch/blob/master/aten/src/TH/generic/THTensorApply.hpp#L7-L10 . `HYPER_TH_OMP_OVERHEAD_THRESHOLD` is used for more fine-grained tasks since HYPER_TH_OMP_OVERHEAD_THRESHOLD = TH_OMP_OVERHEAD_THRESHOLD / 16.


Differential Revision: [D18165752](https://our.internmc.facebook.com/intern/diff/D18165752/)

[ghstack-poisoned]

jianyuh committed:
Update on "[bert] Add the intra-op parallelism for equal operator"


Similar to https://github.com/pytorch/pytorch/pull/28464 and https://github.com/pytorch/pytorch/pull/28477, we would like to enable the intra-op parallelism for layer norm. This will be mapped to the parallel performance win for the BERT/RoBERTa model.

Benchmarking the RoBERTa model with 20 threads:

Before this Diff: P120104857
```
equal                    11.16%           305.851ms        11.16%           305.851ms        4.248ms          NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]     
```


After this Diff:
(grain size is the third parameter when using `at::parallel_for`. As measured below, the performance differences between these different grain size seem to be subtle.)

- grain size = `TH_OMP_OVERHEAD_THRESHOLD`:
```
equal                    1.43%            36.056ms         1.43%            36.056ms         500.783us        NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]
```

- grain size = `HYPER_TH_OMP_OVERHEAD_THRESHOLD`:
```
equal                    1.41%            35.126ms         1.41%            35.126ms         487.855us        NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]
```

- grain size = 1:
```
equal                    1.43%            35.632ms         1.43%            35.632ms         494.886us        NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]
```

Note that the size of  `HYPER_TH_OMP_OVERHEAD_THRESHOLD` and  `TH_OMP_OVERHEAD_THRESHOLD` can be found in https://github.com/pytorch/pytorch/blob/master/aten/src/TH/generic/THTensorApply.hpp#L7-L10 . `HYPER_TH_OMP_OVERHEAD_THRESHOLD` is used for more fine-grained tasks since HYPER_TH_OMP_OVERHEAD_THRESHOLD = TH_OMP_OVERHEAD_THRESHOLD / 16.


Differential Revision: [D18165752](https://our.internmc.facebook.com/intern/diff/D18165752/)

[ghstack-poisoned]

jianyuh committed:
Update on "[bert] Add the intra-op parallelism for equal operator"


Similar to https://github.com/pytorch/pytorch/pull/28464 and https://github.com/pytorch/pytorch/pull/28477, we would like to enable the intra-op parallelism for layer norm. This will be mapped to the parallel performance win for the BERT/RoBERTa model.

Benchmarking the RoBERTa model with 20 threads:

Before this Diff: P120104857
```
equal                    11.16%           305.851ms        11.16%           305.851ms        4.248ms          NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]     
```


After this Diff:
(grain size is the third parameter when using `at::parallel_for`. As measured below, the performance differences between these different grain size seem to be subtle.)

- grain size = `TH_OMP_OVERHEAD_THRESHOLD`:
```
equal                    1.43%            36.056ms         1.43%            36.056ms         500.783us        NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]
```

- grain size = `HYPER_TH_OMP_OVERHEAD_THRESHOLD`:
```
equal                    1.41%            35.126ms         1.41%            35.126ms         487.855us        NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]
```

- grain size = 1:
```
equal                    1.43%            35.632ms         1.43%            35.632ms         494.886us        NaN              0.000us          0.000us          72               [[61, 64, 1024], [61, 64, 1024]]
```

Note that the size of  `HYPER_TH_OMP_OVERHEAD_THRESHOLD` and  `TH_OMP_OVERHEAD_THRESHOLD` can be found in https://github.com/pytorch/pytorch/blob/master/aten/src/TH/generic/THTensorApply.hpp#L7-L10 . `HYPER_TH_OMP_OVERHEAD_THRESHOLD` is used for more fine-grained tasks since HYPER_TH_OMP_OVERHEAD_THRESHOLD = TH_OMP_OVERHEAD_THRESHOLD / 16.


Differential Revision: [D18165752](https://our.internmc.facebook.com/intern/diff/D18165752/)

[ghstack-poisoned]
